File: SessionManager/Models/BitField.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tsunami.Models
{
    public class BitField
    {
        public bool AllSet { get; set; }
        public bool Empty { get; set; }
        public bool NoneSet { get; set; }
        public int NumWords { get; set; }
        public int Size { get; set; }

        public BitField() { /* nothing to do. just for serializator */ }

        public BitField(Core.BitField bf)
        {
            AllSet = bf.all_set;
            Empty = bf.empty;
            NoneSet = bf.none_set;
            NumWords = bf.num_words;
            Size = bf.size;
        }

        //public void ClearBit(int index)
        //{
        //    _bf.clear_bit(index);
        //}

        //public bool GetBit(int index)
        //{
        //    return _bf.get_bit(index);
        //}

        //public bool OpSubscript(int index)
        //{
        //    return _bf.op_Subscript(index);
        //}

        //public void SetBit(int index)
        //{
        //    _bf.set_bit(index);
        //}

    }
}

File: Core/Core/bitfield.cpp
#include "bitfield.h"

using namespace Tsunami::Core;

BitField::BitField()
{
	bitfield_ = new libtorrent::bitfield();
}

BitField::BitField(int bits)
{
	bitfield_ = new libtorrent::bitfield(bits);
}

BitField::BitField(int bits, bool val)
{
	bitfield_ = new libtorrent::bitfield(bits, val);
}


BitField::~BitField()
{
	delete bitfield_;
}

bool BitField::operator[](int index)
{
	return (*bitfield_)[index];
}

bool BitField::get_bit(int index)
{
	return bitfield_->get_bit(index);
}

void BitField::clear_bit(int index)
{
	bitfield_->clear_bit(index);
}

void BitField::set_bit(int index)
{
	bitfield_->set_bit(index);
}

bool BitField::all_set::get()
{
	return bitfield_->all_set();
}

bool BitField::none_set::get()
{
	return bitfield_->none_set();
}

int BitField::size::get()
{
	return bitfield_->size();
}

int BitField::num_words::get()
{
	return bitfield_->num_words();
}

bool  BitField::empty::get()
{
	return bitfield_->empty();
}

BitField::BitField(libtorrent::bitfield & rhs)
{
	bitfield_ = new libtorrent::bitfield(rhs);
}

BitField::BitField(char * b, int bits)
{
	bitfield_ = new libtorrent::bitfield(b, bits);
}

libtorrent::bitfield * BitField::ptr()
{
	return bitfield_;
}
